crates/miden-objects/src/asset/vault.rs

```diff
@@ -261,11 +261,14 @@ impl AssetVault {
 
 impl Serializable for AssetVault {
     fn write_into<W: ByteWriter>(&self, target: &mut W) {
-        // TODO: determine total number of assets in the vault without allocating the vector
-        let assets = self.assets().collect::<Vec<_>>();
+        // Get the number of assets
+        let count = self.asset_tree.num_entries();
 
-        target.write_usize(assets.len());
-        target.write_many(&assets);
+        // Write the count
+        target.write_usize(count);
+
+        // Write all assets using write_many
+        target.write_many(self.assets());
```

Comment:
We can probably get rid of comments here and do something like this:
```rust
target.write_usize(count);
target.write_many(self.assets());
```
I don't think this loses much in terms for readability.